glommio/src/sys/uring.rs

```diff
@@ -297,11 +297,11 @@ where
 
                 let src = peek_source(from_user_data(op.user_data));
                 if let SourceType::Read(PollableStatus::NonPollable(DirectIo::Disabled), slot) =
-                    &mut *src.source_type.borrow_mut()
+                    &mut src.borrow_mut().source_type
```

Comment:
I am surprised this works.
`borrow_mut()` should not survive its scope.
Maybe it works because the scope of this is the whole conditional.

But I guess a safer way would be to write `let mut src_mut = src.borrow_mut()` before the condition so that the reference safely is alive the whole time

Reply:
I believe this is because the lifetime of the member `source_type` is that of the `Ref`/`RefMut` itself. The compiler probably delays the `drop`ping of the borrow based on that. See how it works in a safe context:

```
    pub(crate) fn set_timeout(&mut self, d: Duration) -> Option<Duration> {
        let t = &mut self.inner.borrow_mut().timeout;
        let old = *t;
        *t = Some(TimeSpec64::from(d));
        old.map(Duration::from)
    }
```

Reply:
I have rebased my branch and I no longer use this pattern. I now give the borrow its own, separate lifetime.